src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -1755,206 +1906,299 @@ public static decimal Sum<TSource>(this IEnumerable<TSource> source, Func<TSourc
-                    value = x;
-                    hasValue = true;
-                }
             }
-            if (hasValue) return value;
-            throw Error.NoElements();
+            return value;
         }
 
         public static long? Min(this IEnumerable<long?> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            long? value = null;
-            foreach (long? x in source)
+            Error.CheckSourceNotNull(source);
+            long? value;
+            using (IEnumerator<long?> e = source.GetEnumerator())
             {
-                if (value == null || x < value) value = x;
+                if (!e.MoveNext()) return null;
+                long? cur = e.Current;
+                while (!cur.HasValue)
+                {
+                    if (!e.MoveNext()) return cur;
+                    cur = e.Current;
+                }
+                long valueVal = (value = cur).Value;
+                while (e.MoveNext())
+                {
+                    cur = e.Current;
+                    if (cur.HasValue)
+                    {
+                        long x = cur.Value;
+                        if (x < valueVal)
+                        {
+                            valueVal = x;
+                            value = cur;
+                        }
+                    }
+                }
             }
             return value;
         }
 
         public static float Min(this IEnumerable<float> source)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            float value = 0;
-            bool hasValue = false;
-            foreach (float x in source)
+            Error.CheckSourceNotNull(source);
+            float value;
+            using (IEnumerator<float> e = source.GetEnumerator())
             {
-                if (hasValue)
+                if (!e.MoveNext()) throw Error.NoElements();
+                value = e.Current;
+                while (e.MoveNext())
                 {
+                    float x = e.Current;
+                    if (x < value) value = x;
                     // Normally NaN < anything is false, as is anything < NaN
                     // However, this leads to some irksome outcomes in Min and Max.
                     // If we use those semantics then Min(NaN, 5.0) is NaN, but
                     // Min(5.0, NaN) is 5.0!  To fix this, we impose a total
                     // ordering where NaN is smaller than every value, including
                     // negative infinity.
-                    if (x < value || System.Single.IsNaN(x)) value = x;
-                }
-                else
-                {
-                    value = x;
-                    hasValue = true;
+                    // Not testing for NaN therefore isn't an option, but since we
+                    // can't find a smaller value, we can short-circuit.
+                    else if (float.IsNaN(x)) return x;
```

Comment:
This will change behavior by potentially not enumerating the source to the end where it previously would have (and enumerating the source could have visible side effects).  Something to be discussed.